compiler/rustc_driver_impl/src/lib.rs

```diff
@@ -1195,35 +1192,62 @@ pub fn catch_with_exit_code(f: impl FnOnce() -> interface::Result<()>) -> i32 {
     }
 }
 
-static DEFAULT_HOOK: LazyLock<Box<dyn Fn(&panic::PanicInfo<'_>) + Sync + Send + 'static>> =
-    LazyLock::new(|| {
-        let hook = panic::take_hook();
-        panic::set_hook(Box::new(|info| {
-            // If the error was caused by a broken pipe then this is not a bug.
-            // Write the error and return immediately. See #98700.
-            #[cfg(windows)]
-            if let Some(msg) = info.payload().downcast_ref::<String>() {
-                if msg.starts_with("failed printing to stdout: ") && msg.ends_with("(os error 232)")
-                {
-                    early_error_no_abort(ErrorOutputType::default(), &msg);
-                    return;
-                }
-            };
+/// Stores the default panic hook, from before [`install_ice_hook`] was called.
+static DEFAULT_HOOK: OnceLock<Box<dyn Fn(&panic::PanicInfo<'_>) + Sync + Send + 'static>> =
+    OnceLock::new();
 
-            // Invoke the default handler, which prints the actual panic message and optionally a backtrace
-            // Don't do this for delayed bugs, which already emit their own more useful backtrace.
-            if !info.payload().is::<rustc_errors::DelayedBugPanic>() {
-                (*DEFAULT_HOOK)(info);
+/// Installs a panic hook that will print the ICE message on unexpected panics.
+///
+/// A custom rustc driver can skip calling this to set up a custom ICE hook.
+///
+/// NOTE: this function should not be called more than once per process.
+/// Doing so may lead to infinite recursion or aborts on panic, rather than a normal unwind.
+pub fn install_ice_hook(bug_report_url: Option<&'static str>) {
+    // If the user has not explicitly overridden "RUST_BACKTRACE", then produce
+    // full backtraces. When a compiler ICE happens, we want to gather
+    // as much information as possible to present in the issue opened
+    // by the user. Compiler developers and other rustc users can
+    // opt in to less-verbose backtraces by manually setting "RUST_BACKTRACE"
+    // (e.g. `RUST_BACKTRACE=1`)
+    if std::env::var("RUST_BACKTRACE").is_err() {
+        std::env::set_var("RUST_BACKTRACE", "full");
+    }
 
-                // Separate the output with an empty line
-                eprintln!();
+    let _ = DEFAULT_HOOK.set(panic::take_hook());
+    let bug_report_url = bug_report_url.unwrap_or(
+        "https://github.com/rust-lang/rust/issues/new\
+        ?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md",
+    );
```

Comment:
Can we just use this string on the `install_ice_hook` call-site, instead of using an option?

Reply:
If we do that, the string has to be duplicated between rustc and miri, which seems unfortunate since Miri shouldn't need to care if we change the ICE issue template.

Reply:
oh, I can have a public `DEFAULT_BUG_REPORT_URL` constant in rustc_driver, I'll do that so we can remove the option.